accepted/0000-nohoist.md

```diff
@@ -0,0 +1,299 @@
+    nohoist: ['_all_']
+}
+```
+the file structure will look like this:
+```
+    monorepo
+      |- node_modules/
+        |- workspace-b
+            |- node_modules/
+              |- (... all workspace-b's dependencies...)
+        |- workspace-a
+            |- node_modules/
+              |- (... all workspace-a's dependencies...)
+
+```
+
+# Drawbacks
+
+* Why should we *not* do this?
+
+    - We could argue the current implementation is simple, consistent and correct, therefore, should not introduce nohoist to muddy the water. 
+    - The violating packages, such as react-native, should be corrected to use package manager properly instead of assuming module locations. 
+    - If developers prefer no-hoist, they could just not use "workspaces" or write their own custom scripts to deal with project specific needs.
+
+* Tradeoffs
+    
+    This is a classic simplicity vs usability. yarn can be simple and beautiful but if developers find it hard to use in practice, they will not use it. 
+
+# Alternatives
+
+* What other designs have been considered?
+
+    - There are a few [high level options](https://github.com/yarnpkg/yarn/issues/3882#issuecomment-338478889) discussed in [#3882](https://github.com/yarnpkg/yarn/issues/3882). 
+    - More implementation alternatives are also marked as discussion in their corresponding sections above.  
+    - we could choose not support nohoist-deep-dep(3), as use case (#2) is not very clear to me if we need this or what the expectation is. However, with inherited nohoistList, we get deep-nohoist pretty much for free so decided to just leave it in. We could either not document it publicly or leave it as it is and refine later when needed.
+
+# Unresolved questions
+
+## discussion-1
+**is allowing nohoist in public package bad?**
+
+As mentioned earlier, nohoist only impact yarn hoisting, for other package managers, it's a no-op and should be able to safely ignored... maybe I am missing some problematic use cases, here is what I got:
+
+Let's say a user project A has a dependency on a public package B, which has a dependency on 'react-native' that will need to be excluded from hoist otherwise it won't work. 
+
+||B has workspaces.nohoist=react-native|B didn't have workspaces.nohoist|
+|--|--|--|
+|A doesn't use yarn or workspace|no impact|no impact|
+|A uses yarn workspace|react-native will be excluded from hoisting without A doing anything|A needs to discover then put both B and react-native in its own workspaces.nohoist|
+|A uses yarn workspaces but excluded \_all\_|no impact, none will be hoisted|no impact, none will be hoisted|
+|A uses yarn workspaces but do not want to exclude react-native from hoisting (why?)|conflict! but the public package B is right and react-native should be excluded.|no conflict but B will most likely failed|
+
+In short, having unnecessary nohoist packages might cause inefficiency, but missing a necessary nohoist will lead to compile/execution error.

```

Comment:
That is a good argument.
The counter argument is if we release it now and it goes to public packages then we won't have a chance to change our minds.
Especially if other package managers will implement workspaces.

How about this - we implement it the safer way, `nohoist` only in root but we can implement the nohoist in public packages behind an experimental flag name  `nohoist-yarn-experimental`.
Give it some time and then decide how to proceed.

Reply:
ok, in that case, we can just use the existing `workspaces-experimental` flag, since nohoist is within the workspaces config anyway, probably don't need another flag for nohoist explicitly.

Basically, both package producers and consumers need to have  `workspaces-experimental` flag on in order to access any workspaces config/functionality.